### DISCUSSION TITLE
How to access strorage in a public way?

### DISCUSSION THREAD

**Discussion by zhibisora** [appwrite/appwrite#5068] [Q&A] [answered]
I have a bucket. I hope the images in it can be completely public. It can be accessed by users without login.

I checked many documents, such as sdk, but most of them think that users are logged in and have permissions by default.

I tried to use the link provided by the Web UI, but failed. It can be opened in the browser, but I got an error in my Flutter app.
```
// the link likes this
v1/storage/buckets/{bucketId}/files/{fileId}/view?project={projectId}&mode=admin
```
I know that Flutter sdk should be able to handle these problems, but I hope to be able to use the rest api, but it doesn't work, and I don't know how to add permissions for non-logged-in users according to the rest sdk document or permission document.

Maybe I have neglected something about the public permissions, but I really have no way. I hope I can get help. Thank you!

May be relevant: 
https://github.com/appwrite/appwrite/discussions/3474

**Comment by zhibisora** [comment, marked as answer]
I got it. Because in my settings, the picture inherits the bucket's permission settings, so just change the bucket's permission to any (read only) and you can access it according to the document's instructions. Although, maybe we should add the description of when to add "X-Appwrite-JWT" in headers or the description of the " X-Appwrite-XXX" parameters in the REST document.

**Reply by stnguyen90** [reply]
> Although, maybe we should add the description of when to add "X-Appwrite-JWT" in headers or the description of the " X-Appwrite-XXX" parameters in the REST document.

I'm not sure what you mean 🧐 The headers are documented [here](https://appwrite.io/docs/rest#headers).

**Reply by zhibisora** [reply]
I mean there is no relevant information on this [page](https://appwrite.io/docs/client/storage?sdk=rest-default)

I saw the page you said, but I forgot. Maybe it would be better to add a link on the sdk rest page.

**Comment by zhibisora** [comment]
update: It still not work

![image](https://user-images.githubusercontent.com/73344387/216067563-8c47698f-6737-4509-830a-07d5885e8c54.png)
![image](https://user-images.githubusercontent.com/73344387/216067712-b1576d4b-25f6-47cb-a77a-0ef4eb0579ec.png)
![image](https://user-images.githubusercontent.com/73344387/216068919-b78e586e-5b2c-4982-a8c1-4486dbbde2fe.png)
![image](https://user-images.githubusercontent.com/73344387/216068885-7eea16d3-5132-4d0c-86c9-9794b5a8020a.png)
![image](https://user-images.githubusercontent.com/73344387/216071372-42523cb6-b7ca-4539-ab7a-9c69c0d2116a.png)

Does anyone know what happened? Why is the response 200 but not loaded? I used wildcards to prevent CORS errors and gave storage full permissions, but these are useless.

Browser can access this picture.

Appwrite: 1.2.0
Using flutter-web

**Reply by stnguyen90** [reply]
> ![image](https://user-images.githubusercontent.com/73344387/216068885-7eea16d3-5132-4d0c-86c9-9794b5a8020a.png)

Sometimes, if you don't look at resources right away and try to go back to look at them, you'll get this error.

**Comment by zhibisora** [comment]
![image](https://user-images.githubusercontent.com/73344387/216099246-5b3326f7-7fe4-426a-b031-003c2666d534.png)
It is a CORS error, but the strange thing is that I have set the wildcard CORS hostname. Maybe I should go to the issue to find it.

**Reply by stnguyen90** [reply]
Do you have any sort of reverse proxy in front of Appwrite?

**Reply by zhibisora** [reply]
Yes, I have, it is an nginx server that maps ports to 80 and 443 and provides SSL services.

**Reply by stnguyen90** [reply]
@zhibisora, do you have this directive set?

```
proxy_set_header Host $host;
```

**Reply by zhibisora** [reply]
Yes, I set it.

**Comment by zhibisora** [comment]
I reproduced this CORS error on the web platform.

index.html
```html
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>

<body>
</body>
<script src="https://code.jquery.com/jquery-1.9.1.min.js"></script>
<script>
    $.ajax({
        url: `https://[appwrite hostname]/v1/storage/buckets/[bucket id]/files/[file id]/view`,
        type: 'get',
        headers: {
            // host cannot be set, but it is set up within the url
            // 'Host':'[appwrite hostname]',
            'Content-Type': 'application/json',
            'X-Appwrite-Project':'[project id]'
        },
        success: function (res) {
            console.log(res);
        },
        error: function (e) {
            console.log(e)
        }
    })
</script>

</html>
```
Added 127.0.0.1 as hostname in the project

![image](https://user-images.githubusercontent.com/73344387/216756443-27f91a38-540a-4679-ab5a-e96ae9dbc22a.png)

Start the server in vscode using live server to 127.0.0.1:5500

![image](https://user-images.githubusercontent.com/73344387/216756453-ea4001f3-f7bf-447a-a48e-7b1bd59175f0.png)

Then you can see these in the console

![image](https://user-images.githubusercontent.com/73344387/216756740-a37100a8-daab-4e22-b3ef-8fb327e8fe32.png)


Maybe there's something wrong.

nginx proxy configuration is here (8181 is the port of the appwrite container)

```nginx
#PROXY-START/

location ^~ /
{
    proxy_pass http://127.0.0.1:8181/;
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header REMOTE-HOST $remote_addr;

    add_header X-Cache $upstream_cache_status;

    #Set Nginx Cache
    
    
    set $static_fileuhuRcWi2 0;
    if ( $uri ~* "\.(gif|png|jpg|css|js|woff|woff2)$" )
    {
    	set $static_fileuhuRcWi2 1;
    	expires 12h;
        }
    if ( $static_fileuhuRcWi2 = 0 )
    {
    add_header Cache-Control no-cache;
    }
}

#PROXY-END/
```

**Reply by stnguyen90** [reply]
I wonder if making an ajax request is causing a problem...The typical approach is to put the URL (including the project ID) into an `<img>` element like [this](https://appwrite.io/docs/rest#no-headers).

Would you please try to access your app at http://localhost:5000 instead of http:/127.0.0.1:5000?

Also, it would be best to proxy_pass to appwrite using https rather than http.

Finally, I recommend joining us on [Discord](https://appwrite.io/discord) as troubleshooting over GitHub like this is messy and slow.

**Reply by zhibisora** [reply]
\<img \> is ok. Because it is not an XML request.

localhost is ok, but I don't know why.

It's harder to use https for 127.0.0.1. Maybe I can try it when I have more time.

I will join the discord. I'm not very familiar with it, but thanks for the suggestion.
